src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,371 @@
+                    }
+                    catch (Exception ex)
+                    {
+                        throw new SocksException(SR.net_socks_no_ipv4_address, ex);
+                    }
+
+                    if (addresses.Length == 0)
+                    {
+                        throw new SocksException(SR.net_socks_no_ipv4_address);
+                    }
+
+                    ipv4Address = addresses[0];
+                }
+
+                if (ipv4Address is null)
+                {
+                    Debug.Assert(isVersion4a);
+                    buffer[4] = 0;
+                    buffer[5] = 0;
+                    buffer[6] = 0;
+                    buffer[7] = 255;
+                }
+                else
+                {
+                    ipv4Address.TryWriteBytes(buffer.AsSpan(4), out int bytesWritten);
+                    Debug.Assert(bytesWritten == 4);
+                }
+
+                byte usernameLength = EncodeString(credentials?.UserName, buffer.AsSpan(8), nameof(credentials.UserName));
+                buffer[8 + usernameLength] = 0;
+                int totalLength = 9 + usernameLength;
+
+                if (ipv4Address is null)
+                {
+                    // https://www.openssh.com/txt/socks4a.protocol
+                    byte hostLength = EncodeString(host, buffer.AsSpan(totalLength), nameof(host));
+                    buffer[totalLength + hostLength] = 0;
+                    totalLength += hostLength + 1;
+                }
+
+                await WriteAsync(stream, buffer.AsMemory(0, totalLength), async).ConfigureAwait(false);
+
+                // +----+----+----+----+----+----+----+----+
+                // | VN | CD | DSTPORT |      DSTIP        |
+                // +----+----+----+----+----+----+----+----+
+                //    1    1      2              4
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 8), async).ConfigureAwait(false);
+
+                switch (buffer[1])
+                {
+                    case Socks4_Success:
+                        // Nothing to do
+                        break;
+                    case Socks4_AuthFailed:
+                        throw new SocksException(SR.net_socks_auth_failed);
+                    default:
+                        throw new SocksException(SR.net_socks_connection_failed);
+                }
+                // response address not used
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        private static byte EncodeString(ReadOnlySpan<char> chars, Span<byte> buffer, string parameterName)
+        {
+            try
+            {
+                return checked((byte)Encoding.UTF8.GetBytes(chars, buffer));
+            }
+            catch
+            {
+                Debug.Assert(Encoding.UTF8.GetByteCount(chars) > 255);
```

Comment:
This assert might fail if you pass in invalid Unicode.

Reply:
How do you mean? Afaik the default Encoding.UTF8 doesn't throw but instead replaces chars with `U+FFFD`